chips/cc26x0/src/uart.rs

```diff
@@ -214,6 +231,17 @@ impl kernel::hil::uart::UART for UART {
         });
     }
 
-    #[allow(unused)]
-    fn receive(&self, rx_buffer: &'static mut [u8], rx_len: usize) {}
-}
\ No newline at end of file
+    fn receive(&self, rx_buffer: &'static mut [u8], rx_len: usize) {
+        if rx_len == 0 {
+            return;
+        }
+
+        for i in 0..rx_len {
+            rx_buffer[i] = self.read_byte();
```

Comment:
Reading all the bytes here one at a time is likely a bad choice. The entire system will hang until all the requested bytes are read, which could be an indefinite period of time (for instance, forever if there is a mismatch between expected number of bytes and what actually got received, or several seconds if a human is typing data in through a keyboard).

For the SAM4L platform, we implemented a non-blocking receive interface using the DMA. https://github.com/tock/tock/blob/master/chips/sam4l/src/usart.rs#L853 It starts a reception when commanded, but allows it to run in the background and performs the callback whenever all the data is received. Non-blocking code is certainly a lot more effort to write, but it allows the platform to be more robust because the kernel and applications can continue running while waiting on a reception.

Now, if the goal is to support a bootloader, blocking code might be acceptable. But, in my opinion, I don't think it should be exported through the UART HIL for userland code to use.

Reply:
That totally makes sense. I avoided the use of DMA because I'm having trouble debugging Rust code (it won't let me set breakpoints in gdb). Any advice you'd have there would be appreciated.

I'm also noticing that the power-management-exp branch is like 75 commits ahead of master - should I be developing on that?

Reply:
Also, probably worth pointing out that the `transmit` method is blocking if your message is larger than the FIFO buffer. I attempted to set it up with interrupts (similar to the NRF52 UART), but had trouble getting the TX_FIFO_READY interrupt to flag.